src/main/python/SystemML/defmatrix.py

```diff
@@ -0,0 +1,285 @@
+    dmlOp.dml = [out.ID, ' = ', lhsStr, opStr, rhsStr, '\n']
+    return out
+
+def eval(outputs, outputDF=False, execute=True):
+    """
+    Executes the unevaluated DML script and computes the matrices specified by outputs.
+
+    Parameters
+    ----------
+    outputs: list of matrices
+    outputDF: back the data of matrix as PySpark DataFrame
+    """
+    checkIfMLContextIsSet()
+    reset()
+    matrix.dml = []
+    matrix.script = pydml('')
+    if isinstance(outputs, matrix):
+        outputs = [ outputs ]
+    elif not isinstance(outputs, list):
+        raise TypeError('Incorrect input type')
+    for m in outputs:
+        m.output = True
+        m._visit(execute=execute)
+    if not execute:
+        return ''.join(matrix.dml)
+    matrix.script.scriptString = ''.join(matrix.dml)
+    results = matrix.ml.execute(matrix.script)
+    for m in outputs:
+        if outputDF:
+            m.data = results.getDataFrame(m.ID)
+        else:
+            m.data = results.getNumPyArray(m.ID)
+    
+# Instead of inheriting from np.matrix
+class matrix(object):
+    systemmlVarID = 0
+    dml = []
+    script = None
+    ml = None
+    visited = []
+    def __init__(self, data, op=None):
+        """
+        Constructs a lazy matrix
+        
+        Parameters
+        ----------
+        data: NumPy ndarray, Pandas DataFrame, scipy sparse matrix or PySpark DataFrame. (data cannot be None for external users, 'data=None' is used internally for lazy evaluation).
+        """
+        checkIfMLContextIsSet()
+        self.visited = False 
+        matrix.systemmlVarID += 1
+        self.output = False
+        self.ID = 'mVar' + str(matrix.systemmlVarID)
+        if isinstance(data, SUPPORTED_TYPES):
+            self.data = data
+        elif hasattr(data, '_jdf'):
+            self.data = data
+        elif data is None and op is not None:
+            self.data = None
+            # op refers to the node of Abstract Syntax Tree created internally for lazy evaluation
+            self.op = op
+        else:
+            raise TypeError('Unsupported input type')
+            
+    def eval(self, outputDF=False):
+        eval([self], outputDF=False)
+        
+    def toPandas(self):
+        if self.data is None:
+            self.eval()
+        return convertToPandasDF(self.data)    
+    
+    def toNumPyArray(self):
+        if self.data is None:
+            self.eval()
+        if isinstance(self.data, DataFrame):
+            self.data = self.data.toPandas().as_matrix()
+        # Always keep default format as NumPy array if possible
+        return self.data
+    
+    def toDataFrame(self):
+        if self.data is None:
+            self.eval(outputDF=True)
+        if not isinstance(self.data, DataFrame):
+            if MLResults.sqlContext is None:
+                MLResults.sqlContext = SQLContext(matrix.sc)
+            self.data = sqlContext.createDataFrame(self.toPandas())
+        return self.data
```

Comment:
Yeah if we merge this with the `Matrix` class we could merge this method with `toDF()` using the later name to be consistent with Spark.